milter-size.8
.\" Copyright (C) 2003-2007  Richard Gooch
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" Richard Gooch may be reached at  http://www.safe-mbox.com/~rgooch/
.\"
.\"	milter-size.8		Richard Gooch	7-JUL-2007
.\"
.TH MILTER-SIZE 8 "7 Jul 2007" "GoochUtils"

.SH NAME
milter-size \- sendmail milter to reject certain message sizes

.SH SYNOPSIS
.B milter-size
.B -l low
.B -h high
.B -p port
.B [ -r remhost ]
.B [ -R remport ]

.SH DESCRIPTION
The \fImilter-size\fP programme is a sendmail milter (mail filter)
which can be used to reject messages of certain sizes during the SMTP
transaction. It is a little more flexibility than the sendmail
message size limit. It provides a simple and quick way to block
rampaging virus attacks (where a very large number of unwanted
messages of nearly the same size flood your email system) but still
allowing the administrator to limit the negative effects of the
milter.

The milter will fork a background process which should never exit
(unless there is some internal failure).

Rejected emails will result in a 55x code being returned by sendmail,
with the following status string:
.P
"body length: %d rejected by administrator"
.P
where "%d" is the number of bytes in the message body.

.SH OPTIONS
.TP
.B \-l low
The lower limit on message body size. Messages above this limit are
passed. The default is 0.
.TP
.B \-h high
The higher limit on message body size. Messages below this limit are
passed. The default is infinite.

If \fIlow\fP is less than \fIhigh\fP then messages equal or larger
than \fIlow\fP AND smaller or equal to \fIhigh\fP are passed, all others are
rejected (thus allowing only messages within a specified size range).

If \fIlow\fP is greater than \fIhigh\fP then messages larger than
\fIlow\fP OR smaller than \fIhigh\fP are passed, all others are
rejected (thus rejecting messages within a specified size range).
.TP
.B \-p port
The port to listen on for connections from sendmail.
.TP
.B \-r remhost
If specified, the size of each message is sent via a UDP packet to
\fIremhost\fP. The packet will contain a single, network-endian
(i.e. big endian) 32 bit unsigned integer which is the size in bytes
of the message body. Note that the size is sent irrespective of
whether \fImilter-size\fP rejects the message.
.TP
.B \-r remport
UDP size messages are sent to \fIremport\fP. The default is 1024.

.SH CAVEATS
The message body usually includes carriage returns, as these are
present in the SMTP transaction. This should be taken into account
when setting message body size limits.

.SH EXAMPLES
The SoBig virus has message body sizes ranging from 99 KB to 105 KB.
Blocking all messages over 99 KB is quite drastic, as there may be
many legitimate emails which are quite large. Blocking only messages
between 99 KB and 105 KB reduces the effect on legitimate email, but
still blocks the virus. The following command implements this:

milter-size -l 105000 -h 99000 -p local:/var/run/milter-size.sock

If you want to block messages over 1 MB, do this:

milter-size -h 1000000 -p local:/var/run/milter-size.sock

If you just want to log the size of message bodies, do this:

milter-size -r myhost.mydomain -p local:/var/run/milter-size.sock

.SH SENDMAIL CONFIGURATION
You should consult the sendmail documentation for information on how
to configure a milter, but here is a suggestion for your M4
configuration file:

INPUT_MAIL_FILTER(`milter-size', `S=local:/var/run/milter-size.sock')

.SH NOTE
The sendmail configuration file will need to be updated so that this
milter is used, and sendmail will need to be restarted.
.SH AUTHOR
Richard Gooch http://www.safe-mbox.com/~rgooch/

.SH AVAILABILITY
The \fBmilter-size\fP programme is available from:
ftp://ftp.atnf.csiro.au/pub/people/rgooch/utilities/


milter-size.c
/*  sizemilter.c

    Main file for  sizemilter  (Sendmail Milter to control message sizes).

    Copyright (C) 2003  Richard Gooch

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Richard Gooch may be reached at  http://www.safe-mbox.com/~rgooch/
*/

/*
    This milter allows you to reject messages of a specified size range.


    Written by      Richard Gooch   29-AUG-2003

    Updated by      Richard Gooch   31-AUG-2003: Improved resource management.

    Last updated by Richard Gooch   11-SEP-2003: Added size reporting via UDP.


*/
#ifndef _REENTRANT
#  error  Compile with _REENTRANT
#endif

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <netdb.h>
#include <libmilter/mfapi.h>
#include <errno.h>

#define ERRSTRING strerror (errno)


struct private_struct
{
    unsigned int  nbytes;    /*  Number of bytes in the message body  */
};

static unsigned int minbytes = 0;
static unsigned int maxbytes = 0;
static int remhost_fd = -1;


static sfsistat mlfi_connect (SMFICTX *ctx, char *hostname,
			      _SOCK_ADDR *hostaddr)
{
    struct private_struct *priv;

    /*  Allocate per-connection private memory  */
    if ( ( priv = malloc (sizeof *priv) ) == NULL )
    {
	/*  Can't accept this message right now  */
	return (SMFIS_TEMPFAIL);
    }
    smfi_setpriv (ctx, priv);
    return (SMFIS_CONTINUE);
}   /*  End Function mlfi_connect  */

static sfsistat mlfi_envfrom (SMFICTX *ctx, char **envfrom)
{
    struct private_struct *priv = smfi_getpriv (ctx);

    /*  Initialise private memory for this new message  */
    memset (priv, 0, sizeof *priv);
    return (SMFIS_CONTINUE);
}   /*  End Function mlfi_envfrom  */

static sfsistat mlfi_body (SMFICTX *ctx, u_char *bodyp, size_t bodylen)
{
    struct private_struct *priv = smfi_getpriv (ctx);

    priv->nbytes += bodylen;
    return (SMFIS_CONTINUE);
}   /*  End Function mlfi_body  */

static sfsistat mlfi_eom (SMFICTX *ctx)
{
    struct private_struct *priv = smfi_getpriv (ctx);
    char txt[256];

    if (remhost_fd > -1)
    {   /*  Transmit the size for analysis  */
	uint32_t buf = htonl (priv->nbytes);

	if ( (write (remhost_fd, &buf, 4) < 4) && (errno == ECONNREFUSED) )
	    write (remhost_fd, &buf, 4);  /*  Previous call consumed error  */
    }
    if ( (minbytes == 0) && (maxbytes == 0) ) return (SMFIS_CONTINUE);
    if (minbytes <= maxbytes)
    {   /*  Legal message: min <= size <= max  */
	if ( (priv->nbytes >= minbytes) && (priv->nbytes <= maxbytes) )
	    return (SMFIS_CONTINUE);
    }
    else
    {   /*  Illegal message: max <= size <= min  */
	if ( (priv->nbytes < maxbytes) || (priv->nbytes > minbytes) )
	    return (SMFIS_CONTINUE);
    }
    sprintf (txt, "body length: %d rejected by administrator", priv->nbytes);
    smfi_setreply (ctx, "551", NULL, txt);
    return (SMFIS_REJECT);
}   /*  End Function mlfi_eom  */

sfsistat mlfi_close (SMFICTX *ctx)
{
    struct private_struct *priv = smfi_getpriv (ctx);

    smfi_setpriv (ctx, NULL);
    if (priv) free (priv);
    return (SMFIS_CONTINUE);
}   /*  End Function mlfi_close  */

static struct smfiDesc smfilter =
{
    "SizeMilter",	/* filter name */
    SMFI_VERSION,	/* version code -- do not change */
    SMFIF_ADDHDRS,	/* flags */
    mlfi_connect,	/* connection info filter */
    NULL,		/* SMTP HELO command filter */
    mlfi_envfrom,	/* envelope sender filter */
    NULL,		/* envelope recipient filter */
    NULL,		/* header filter */
    NULL,		/* end of header */
    mlfi_body,		/* body block filter */
    mlfi_eom,		/* end of message */
    NULL,		/* message aborted */
    mlfi_close		/* connection cleanup */
};

int main (int argc, char *argv[])
{
    int c;
    int remport = 1024;
    char *args = "l:h:p:r:R:";
    char *remhost = NULL;

    switch ( fork () )
    {
      case 0:   /*  Child   */
	break;
      case -1:  /*  Error   */
	fprintf (stderr, "Error forking\n");
	exit (EX_UNAVAILABLE);
	/*break;*/
      default:  /*  Parent  */
	_exit (0);
	/*break;*/
    }
    /* Process command line options */
    while ( ( c = getopt (argc, argv, args) ) != -1 )
    {
	switch (c)
	{
	  case 'l':
	    if ( !optarg || (*optarg == '\0') )
	    {
		fprintf (stderr, "Illegal minimum: %s\n", optarg);
		exit (EX_USAGE);
	    }
	    minbytes = atoi (optarg);
	    break;
	  case 'h':
	    if ( !optarg || (*optarg == '\0') )
	    {
		fprintf (stderr, "Illegal maximum: %s\n", optarg);
		exit (EX_USAGE);
	    }
	    maxbytes = atoi (optarg);
	    break;
	  case 'p':
	    if ( !optarg || (*optarg == '\0') )
	    {
		fprintf (stderr, "Illegal conn: %s\n", optarg);
		exit (EX_USAGE);
	    }
	    smfi_setconn (optarg);
	    break;
	  case 'r':
	    if ( !optarg || (*optarg == '\0') )
	    {
		fprintf (stderr, "Illegal remhost: %s\n", optarg);
		exit (EX_USAGE);
	    }
	    remhost = optarg;
	    break;
	  case 'R':
	    if ( !optarg || (*optarg == '\0') )
	    {
		fprintf (stderr, "Illegal remport: %s\n", optarg);
		exit (EX_USAGE);
	    }
	    remport = atoi (optarg);
	    break;
	}
    }
    if (remhost)
    {
	struct sockaddr_in in_addr;
	struct hostent *hostent;

	if ( isascii (remhost[0]) && isdigit (remhost[0]) )
	{
	    /*  Numeric Internet address  */
	    if ( !inet_aton (remhost, &in_addr.sin_addr) )
	    {
		fprintf (stderr, "Invalid host address: \"%s\"\n", remhost);
		exit (EX_USAGE);
	    }
	}
	else if ( ( hostent = gethostbyname (remhost) ) == NULL )
	{
	    fprintf (stderr, "Error looking up host: \"%s\"\t%s\n",
		     remhost, ERRSTRING);
	    exit (EX_OSERR);
	}
	else
	{
	    if (hostent->h_length != 4)
	    {
		fprintf (stderr, "Hostlength: %d is not 4\n",
			 hostent->h_length);
		exit (EX_OSERR);
	    }
	    in_addr.sin_addr.s_addr = *(uint32_t *) hostent->h_addr_list[0];
	}
	in_addr.sin_family = AF_INET;
	in_addr.sin_port = htons (remport);
	if ( ( remhost_fd = socket (AF_INET, SOCK_DGRAM, 0) ) < 0 )
	{
	    fprintf (stderr, "Error creating socket\t%s\n", ERRSTRING);
	    exit (EX_OSERR);
	}
	if (connect (remhost_fd, (struct sockaddr *) &in_addr, sizeof in_addr)
	    != 0)
	{
	    fprintf (stderr, "Error connecting\t%s\n", ERRSTRING);
	    exit (EX_OSERR);
	}
    }
    if (smfi_register (smfilter) == MI_FAILURE)
    {
	fprintf (stderr, "smfi_register failed\n");
	exit (EX_UNAVAILABLE);
    }
    close (0);
    close (1);
    close (2);
    return smfi_main ();
}   /*  End Function main  */
